Customise Homepage

sheralim012 opened the pull request:
Fixed #75 

* Added featured content support on Homepage in wagtail admin
* Added banners support on Homepage wagtail admin
* Made menu flat (no sub-navigation)
* Set default to `True` for **show in menus** for sections and articles

sheralim012 commented:
Hey @dannyparsons

<img width="1620" alt="Screenshot 2021-05-30 at 8 40 22 PM" src="https://user-images.githubusercontent.com/49383675/120112637-16637400-c190-11eb-8846-0fc513d82745.png">
menu support included, you can create a new menu from Settings>Main Menu, just make sure the page you are adding is eligible to be shown in the menu by checking the promote tab. Levels for the menu could also be selected from menu settings.

<img width="300" alt="Screenshot 2021-05-30 at 8 40 55 PM" src="https://user-images.githubusercontent.com/49383675/120112820-f7191680-c190-11eb-8cb5-b48a2a3ec7f7.png">
featured content support included, you can add featured contents on homepage.

<img width="200" height="300" alt="Screenshot 2021-05-30 at 9 09 36 PM" src="https://user-images.githubusercontent.com/49383675/120112760-9a1d6080-c190-11eb-8911-e7c879bf12d8.png">
footer support included, you can create new snippet for footer from wagtail menu.

dannyparsons commented:
Thanks for this. This is looking good, a few points:

- Can we set **Show in menus** to be checked by default for articles and sections? I can see this being confusing for CMS users when creating the menu and not knowing why an item doesn't appear.
- In Settings > Main Menu > Settings, can you set the default to be **1. No Sub-navigation (flat)** - this is what we want the sites to use.
- The Featured Content works well. When I run it, it still shows all the articles first and then the featured content below. We only want the featured content to show now, not all the articles. Not sure if that's just for me?
- Footers also display. Probably need a design and functionality discussion around these, but not for this PR

sheralim012 commented:
<img width="1544" alt="Screenshot 2021-06-01 at 1 12 05 PM" src="https://user-images.githubusercontent.com/49383675/120300533-7586cd00-c2e5-11eb-8aa4-eb10a74d81ba.png">
* banners included, with clickable images for internal and external pages
* made **Show in menus** default to checked
* made menu sub navigation to **flat**
* removed old articles from homepage

dannyparsons commented:
Thanks. Can you swap Banners and Featured Content sections in the CMS when editing the homepage? Banners appear above so it will be more logical for it to be above in the CMS as well.

sheralim012 commented:
unified migration files, so before pulling new code run the following commands
```
make ssh
cd /app/
python manage.py migrate home 0004
```

then you can pull new code and run new migrations

dannyparsons commented:
Looks good. Minor points:

- Could you make the Banner image mandatory? It isn't in v1 but I think it needs to be otherwise there isn't anything to display
- Could you add the help text on hover for the banner fields? (see below) Something like:
Banner image: Image to display as the banner
Link page: Optional page to which the banner will link to
External link: Optional external link which a banner will link to eg https://www.google.com

![image](https://user-images.githubusercontent.com/11214914/120352054-4f1e5d00-c2f8-11eb-8b5a-1e163fcc90b2.png)

sheralim012 commented:
@dannyparsons addressed both points.

dannyparsons commented:
Thanks. 

@istride raised some important points on use of banners https://github.com/unicef/iogt/issues/75#issuecomment-852264394 I think that needs to be talked through with him next.

sheralim012 commented:
@dannyparsons banner and footer are implemented as pages now, kindly check.

sheralim012 commented:
fixed migration files, so before pulling new code run the following commands
```
make ssh
cd /app/
python manage.py migrate home 0004
```
then you can pull new code and run new migrations

dannyparsons commented:
Thanks.

I was assuming that banners and footers would still be selected in a field from the homepage like the featured content. This would be desirable for the CMS editors. It would mean easier control over which banners are included and in what order. Banners could be created under a Banners Index Page (like in v1), instead of having to be children of the homepage. The same for footer. I think that would make it CMS editing clearer.

sheralim012 commented:
Hey @dannyparsons,

I have implemented Banner and Footer Index Pages (to hold all banners and footers respectively) and Banner and Footer for individual entity just like V1 and I have also added the ability to select banners for the home page just like featured content. One thing I want to confirm is that should selecting footer also be an option on home page?

sheralim012 commented:
fixed migration files, so before pulling new code run the following commands
```
make ssh
cd /app/
python manage.py migrate home 0004
```
then you can pull new code and run new migrations

dannyparsons commented:
Thanks. I like this implementation because:

- Banners are defined in their own place (Banner Index Page) so don't clutter the site structure. 
- The selection of banners as a field on the homepage means the things that will appear on the homepage are defined in one place. Also Laurie mentioned that the ordering of banners on the Banner Index Page in v1 is quite confusing. Ordering them on the homepage field is a much nice interface.
- Banners don't have to be put back into draft for them to be hidden (they are just removed from the homepage)
- If we ever wanted to use banners on other pages we could easily do this.

Footers are maybe different because they don't just appear on the homepage they appear across the site. So conceptually they don't belong on the homepage. But it would be much better to have the ordering of them done through the page selection way instead of the Footer Index Page.

I think we need input from @istride before deciding.

sheralim012 commented:
@dannyparsons @istride reverted footer changes and currently, this PR contains featured content and banner. 

Also fixed migration files, so before pulling new code run the following commands
```
make ssh
cd /app/
python manage.py migrate home 0004
```
then you can pull new code and run new migrations

istride commented on home/templates/home/tags/banners_list.html:
```diff
@@ -0,0 +1,21 @@
+              {% elif banner.specific.external_link %}
+                <a href="{{ banner.specific.external_link }}" target="_blank">
+              {% else %}
+                <a href="#"></a>
```
I think the anchor tag should not be closed here.

sheralim012 replied:
removed unnecessary closing tag

sheralim012 commented on home/models.py:
```diff
@@ -15,12 +17,41 @@
+
     def get_context(self, request):
         context = super().get_context(request)
         context['articles'] = self.get_descendants().type(Article)
```
done

sheralim012 commented on home/models.py:
```diff
@@ -15,12 +17,41 @@
     def get_context(self, request):
         context = super().get_context(request)
         context['articles'] = self.get_descendants().type(Article)
+        context['banners'] = [home_page_banner.banner_page for home_page_banner in self.home_page_banners.all()]
```
fetching only live content now

sheralim012 commented:
@istride changes addressed.

istride commented:
Merged manually as I had to merge changes from develop and recreate the database migrations.
